### DISCUSSION TITLE
Cross Controller Support

### DISCUSSION THREAD

**Discussion by m88youngling** [LBPUnion/ProjectLighthouse#429] [Ideas]
Now, I know what you're thinking. Cross controller doesn't need servers to work, right? It works just fine offline? Well, I've been asking around, and it seems like nobody can get it to work, regardless of software version, hardware, or Internet. After a lot of trial and error, this struck me as odd. I then decided to take Ritchie's advice, which was to investigate whether or not Cross Controller actually relies on the official servers. And guess what, it looks like he was right! 

![image](https://user-images.githubusercontent.com/89745384/184468196-be1e2042-1e7d-47e2-928c-4b3c8cc29c1d.png)

This is a screenshot from a packet capture I recorded from the PlayStation Vita Cross Controller app as it tries to connect to a PS3 (turned off in this example). However, most notably, it tries to connect to the official Vita server addresses via DNS. Very peculiar for an 'offline capable' feature. 

This gives me a few ideas, 

- It might be possible to experiment with responding to a Cross Controller client to see what it does next. I'm not sure if any packet captures exist of this feature working when the servers were online. 
- According to Wireshark, the protocol used for these communications is DNS. It might be possible to run Cross Controller features on an OFW Vita by changing DNS settings. However, for users who are already interested in running LBP Vita on CFW/HFW for Lighthouse, modifying the Cross Controller app might be easier. 

https://drive.google.com/file/d/1N68z_oBpjXQdDri6CKNA3E9iBIQO5t0Y/view?usp=sharing - Packet Capture

**Comment by jvyden** [comment]
I will take a look at these packet captures. Very interesting find.

**Reply by jvyden** [reply]
Hmm, yeah these packet captures are vague. The Vita seems to not like the response it's getting from the DNS server; it executes the same query multiple times, but there are no errors in the response.

The one thing that does stand out to me is the fact that the request wants a recursive query, but the server will not send them. This is likely why it's not accepting it.

Interestingly I cant `nslookup` to give me a valid query response. That's probably because I'm using Windows 7.

**Reply by jvyden** [reply]
Nevermind, a record is there but it doesn't seem to have an address attached.

**Reply by m88youngling** [reply]
If it helps, this section of the capture occurs after generating a failure message. You may have figured that out already but as long as the message stays on screen, it keeps generating these lines over and over 

![image](https://user-images.githubusercontent.com/89745384/184468871-9363bda3-0036-4724-adeb-3a0270f2750e.png)

I'm guessing this is just it looking at me with this emoji 😐

**Reply by jvyden** [reply]
![image](https://user-images.githubusercontent.com/40577357/184468896-bbdba357-f9ae-4e80-aef0-ca5a27f57b1a.png)

**Reply by jvyden** [reply]
> If it helps, this section of the capture occurs after generating a failure message. You may have figured that out already but as long as the message stays on screen, it keeps generating these lines over and over
> 
> ![image](https://user-images.githubusercontent.com/89745384/184468871-9363bda3-0036-4724-adeb-3a0270f2750e.png)
> 
> I'm guessing this is just it looking at me with this emoji neutral_face

This is normal network traffic and can be ignored.

**Comment by ZippyZolton** [comment]
you're*

**Comment by jvyden** [comment]
Yeah, the reason it's failing is because of a bad DNS query. There is a record, but the record has no IP addresses. There's a (small) chance it will work if we override DNS, but it's probably better to try patching the EBOOT to see what happens.

**Reply by jvyden** [reply]
It really depends on what's easier. I looked it up and tried to find a hosts file for Vita's OS (a file on every modern operating system to help override DNS queries) but could not find one.

**Comment by jvyden** [comment]
@m88youngling I noticed you're actually using your router for DNS, perhaps you can override a domain in your router's DNS settings as a quick and dirty hack?

**Reply by m88youngling** [reply]
Give me some instructions and I could try it?

**Reply by jvyden** [reply]
What router do you have?

**Reply by m88youngling** [reply]
Google Nest Wi-Fi :trollface:

**Reply by m88youngling** [reply]
I manage all the settings in the app

**Reply by jvyden** [reply]
Okay, this isn't what I thought. The router is simply acting as a proxy; not actually providing a real DNS server. I'll look up how to set up a quick DNS server

**Reply by jvyden** [reply]
Yeah, it looks easier to just patch the EBOOT.

**Reply by jvyden** [reply]
@m88youngling Try patching it to the production server like you would LBP Vita, except target cross controller. Lets see what happens.

**Reply by m88youngling** [reply]
Okay, I will teach myself to do it lol might take me a bit because im in a sticky situation with my Vita CFW, so I'll have to figure that out first

**Reply by jvyden** [reply]
Worst case scenario Ture can help; he should be getting up not too long in the future.

**Reply by m88youngling** [reply]
![image](https://user-images.githubusercontent.com/89745384/184471705-de726da1-6f96-4e38-a78c-9fe23f89972f.png)

I decrypted the EBOOT and found what I think is the server address. It looks like there's more than one.

**Reply by jvyden** [reply]
I believe all of those are already in UnionPatcher. You should be able to patch this just fine.

**Reply by m88youngling** [reply]
IT WORKED!!@!!!

**Reply by jvyden** [reply]
lmao was that seriously it just patch the eboot?

**Comment by jvyden** [comment]
Okay, to summarize:

- LBP Cross controller depends on the old PS Vita servers
- The DNS entries for the old servers no longer have addresses, causing LBP Cross Controller to fail to connect
- Patching the cross controller app's EBOOT will resolve the issue
- No changes to Lighthouse or UnionPatcher were required
- Cross Controller does not actually use the server; it just simply wants to know that the server exists. Either that or this is a leftover of LBPV.

**Reply by m88youngling** [reply]
Can confirm, it works beautifully. You don't even need to be connected to PSN on either console. I played the first level of the Cross Controller DLC pack without any issues

**Reply by jvyden** [reply]
Can you upload a packet capture of you connecting with cross-controller? I'm curious as to what it uses the server for.

**Reply by m88youngling** [reply]
I will try and send this to you when I am conscious LOL I'm so tired

**Reply by m88youngling** [reply]
So, I was unable to get my PS3 to connect to my laptop via hotspot, even with 2.4 GHz. It's probably some kind of security issue. This means I won't be able to have both my Vita and my PS3 on the same network, so I won't yet be able to see what the traffic looks like during normal play. 

However, I made another packet capture of the Vita after patching the Cross Controller app, and it results in something slightly different. As far as I can tell, it looks like it just checks and sees if the server is there, and then continues to attempt to connect to the PS3. 
![image](https://user-images.githubusercontent.com/89745384/184515811-5d353f64-de5f-453c-a99c-69acd600752a.png)

If that's true, then it may not be using Lighthouse for anything at all. Arcadius suggested that it might just be leftover code from LBP Vita where the Cross Controller app is based off of.
[Lighthouse Cross Controller Partial.zip](https://github.com/LBPUnion/ProjectLighthouse/files/9332383/Lighthouse.Cross.Controller.Partial.zip)

**Reply by jvyden** [reply]
Yeah, it just looks like it wants to know that the server is there. I don't see any communication with the server in this packet capture.

**Reply by TorutheRedFox** [reply]
it might just be to make damn sure that the console has internet access by sending a known good (not anymore lol) DNS query

**Reply by darthHunterous** [reply]
Do you guys think there's a way to overcome this issue without patching the EBOOT? Having an OFW way would be nice, as it would be deemed legit in some PlayStation gaming communities. For example, in Saints Row 3 you can re-route the offline server for character upload to anything else and it will make the feature somewhat accessible. Would something similar work with the Cross Controller?
https://forum.psnprofiles.com/topic/66339-how-to-get-the-jumped-in-trophy-after-the-servers-shutdown/

**Reply by m88youngling** [reply]
Cross controller probably yes. And probably to *any* website or server but that hasn't been tested. We think cross controller is just doing a network check. That's all. If you're curious about other parts of Lighthouse, DNS will not work, see this thread https://www.lbpunion.com/forum/union-hall/project-lighthouse-faq-littlebigplanet-private-servers-megathread

**Comment by TorutheRedFox** [comment]
doesn't cross controller also need slot filtering and a separate earth decoration on the PS3 side?

**Reply by m88youngling** [reply]
I will check compatibility by trying to upload a cross controller level soon

**Reply by m88youngling** [reply]
Hello again,  I did test this. Lighthouse does not currently support having a cross controller earth and moon uploaded to the server. I _was_ able to upload a level, but I didn't notice any particular cross controller filtering at all. The level is playable by other users who don't have cross controller.

I'm actually not sure why cross controller uses a separate earth and moon. I'm not sure if these elements are even visible to anyone other than the user with cross controller. So, in that case, I don't think supporting the feature is that important. It may actually be detrimental to the server to support it if it has no real reason to be on the server.

I have however experienced more rate limiting I think with the server, suggesting that maybe there are various requests going on that the server thinks are too frequent. I haven't tested this with an extended session without cross controller yet, it's possible that issue is completely independent of CC.

**Reply by TorutheRedFox** [reply]
from what I remember, the cross controller mode was made to be completely separate from the main game, giving you a fresh moon and fresh earth that you can decorate over again to keep cross controller levels separate from normal ones while not having either detract from the other's slot limits

**Reply by m88youngling** [reply]
I see! That sounds right to me. For now though it looks like things are functional which is great, I feel like we got really lucky for the most part, even if the cross controller earth and moon aren't supported

**Reply by TorutheRedFox** [reply]
filtering peripheral based levels is generally unimplemented

so no cross controller or move pack filtering

**Reply by m88youngling** [reply]
I'd say move pack filtering would be more important to support than cross controller if it comes down to prioritization, given that there are significantly more move controller users if I had to guess

I'm sure both will be supported eventually

**Reply by TorutheRedFox** [reply]
move pack filtering tbh is less important as it's really just an optional filter you can apply in the ingame sort menu

**Reply by Ezoiar** [reply]
I don't think there was any move pack separation at the official servers, but filtered search feature would work best if you want move pack filter.

**Reply by TorutheRedFox** [reply]
the move pack levels were treated as their own game by the game filter

**Comment by m88youngling** [comment]
Starting a new comment on this thread to say that thanks to ImBasicallyAlex's suggestion on Discord and testing by Vilijur, we've discovered more specifically what needs to be changed in the EBOOT to allow cross controller to work correctly. The original line is shown below, 

![image](https://user-images.githubusercontent.com/89745384/186809706-099e4f98-f57a-4dc7-a3ae-a31ab4f54e44.png)

All you have to do is change the url to the Lighthouse in the same format. Include .sessionmaster. They suspect that it uses the server for something more than that, but this hasn't been experimentally verified. I've experienced rate limiting and a lot of authentication attempts when using cross controller, but I've asked the USC team to report bugs they find. Their EBOOTs will be different than mine since they didn't patch the other URLs in the file. Who knows what their performance will be like until they report it. I'll continue to drop my findings in here, as it might help with future feature implementation.

**Reply by vilijur** [reply]
To add onto this, there's probably a way to integrate this patching into [UnionPatcher](https://github.com/LBPUnion/UnionPatcher) to allow people to patch their app without having to hex edit, which could make the process more simple.

**Reply by vilijur** [reply]
I've done quite a bit of experimenting since then, and have came back with the following conclusions...
In order to successfully patch a Cross-Controller eboot, the URL must:
- Be short enough to fit in the eboot without adding any extra bytes to the file (I have not tested to see if there needs to be space between the URL and the `true`, or if the `true` is even necessary) (also any extra or missing bytes will make the eboot unbootable)
- Be typed at **the beginning of the original URL**
- End with `.sessionMaster`
- I tried a URL with just an image and that didn't work so it's probably gotta have some sorta html to it idk I'm not a web developer

Also, as one last tidbit, I have not yet tested the DNS theory, mainly because I have no clue what I'm doing when it comes to that stuff. Theoretically speaking, I still think it would work, but without practice, it's hard to tell.
anyway that's all from me, time to not respond to anything else apart from this discussion because cross-controller is my life now lol
ok bye

**Comment by ShadowMan2002** [comment]
I really don't know much about game programming but I spent a while looking at the original Eboot.bin and the one modified after the LBP Union and what I noticed, apart from the fact that the patcher makes a cut to the original file, I realized that the eboot .bin apart from an HTTP or HTTPS address, there is a port number that is 10060, looking for the port address plus the web address, yes, it is down but I tried to access the port through VPN and the port does not work either.

Already half out of ideas, I tried to see what I could try and it would be curious, we have tried to return the remote function to something that we must change apart from a web address, we must have a port, but we did not think about the USB port, the PSvita USB Port sends information to the PS3 constantly as from Host to Server, applications like Vita Stick take advantage of this to become a PS3 or PC control via USB, you could try making a tool that allows you to simulate the Cross-Controller through USB or through the connection Bluetooth of the system, it is an idea, since it is like we are at a dead point, since without the port and without the address it is somewhat difficult, unless someone gets the domain and reactivates it from within or finds a way to restore through the Project Lighthouse page

**Reply by jackcaver** [reply]
Restoring cross controller app is as simple as modifying hosts file on your PC, you just need to redirect the domain that the game checks for to something like `127.0.0.1`, the app doesn't seem to even try to connect to it, it just checks if the domain exists, so in theory LBPUnion could just host a custom DNS server for that purpose, but for now if you want a solution that works on OFW, just host a WiFi hotspot from a device where you could modify the `hosts` file and that's it, although i only tested it on windows, so i'm not sure if it works on Linux or MacOS
